### PR TITLE
first idm metadata test

### PR DESCRIPTION
**Pull request by markg58**
Jira issue? CLOUD-1078
Release 6.5.0 backport required? No
6.5.0 doc changes needed? No
7.0.0 doc changes needed? No
Required README updates made? No

### CODE REVIEW

**Comment by markg58** [comment]
Note:  test fails because of **Expected IDM build revision 9c368d2, but found unknown**

File: samples/config/smoke-deployment/env.sh [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # Environment settings for the deployment
 
 # k8s namespace to deploy in
-NAMESPACE=smoke
+NAMESPACE=markg
```
Comment (andrewvinall) [inline samples/config/smoke-deployment/env.sh]: Namespace should be `smoke`

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        # resource_version = 'resource_version_example' # str | When specified with a watch call, shows changes that occur after that particular version of a resource. Defaults to changes from the beginning of history. When specified for list: - if unset, then the result is returned from remote storage based on quorum-read flag; - if it's 0, then we simply return what we currently have in cache, no guarantee; - if set to non zero, then the result is at least as fresh as given rv. (optional)
+        timeout_seconds = 56 # int | Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity. (optional)
+
+        try:
+            api_response = v2.list_namespaced_stateful_set(namespace,
+                                                                     include_uninitialized=include_uninitialized,
+                                                                     pretty=pretty, timeout_seconds=timeout_seconds)
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: None standard formatting.

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+# Lib imports
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: Some of these imports are unused.
Also some are commented out.

Comment (andrewvinall) [inline cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: We need
* a copyright header
* file description

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+from pprint import pprint
+
+# Framework imports
+# from ProductConfig import IDMConfig
+# from utils import logger, rest
+
+
+class TestMilestoneValidation(object):
+
+    def list_idm_objects(self):
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: This method could be private.

File: cicd/forgeops-tests/lib/utils/idm_pod.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+# Framework imports
+from ProductConfig import IDMConfig
+from utils import logger, rest, kubectl
+from utils.pod import Pod
+
+
+class IDMPod(Pod):
+    PRODUCT_TYPE = 'openidm'
+    ROOT = os.path.join(os.sep, 'opt', 'openidm')
+#    TEMP = os.path.join(ROOT, 'fr-tmp')
```
Comment (andrewvinall) [inline cicd/forgeops-tests/lib/utils/idm_pod.py]: Commented out code.

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        # Configs can be set in Configuration class directly or using helper utility
+        config.load_kube_config()
+
+        v2 = client.AppsV1beta2Api()
+
+        # create an instance of the API class
+        namespace = 'markg'
+        include_uninitialized = True
+        pretty = True
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: What is the purpose of this block of commented out code?

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+from kubernetes.client.rest import ApiException
+from kubernetes.stream import stream
+from pprint import pprint
+
+# Framework imports
+# from ProductConfig import IDMConfig
+# from utils import logger, rest
+
+
+class TestMilestoneValidation(object):
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: I am unsure of the purpose of this class as there is an overlap with the metadata tests.
Also, the tests do not assert anything.
The file needs docstrings.

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                                                                     pretty=pretty, timeout_seconds=timeout_seconds)
+            pprint(api_response)
+        except ApiException as e:
+            print("Exception when calling AppsV1beta2Api->list_namespaced_stateful_set: %s\n" % e)
+
+    def test_idm_objects(self):
+        self.list_idm_objects()
+
+    def test_list_pods(self):
+        # Load the kubeconfig file from $HOME/.kube/config
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: Are these comments necessary?

File: cicd/forgeops-tests/lib/utils/kubectl.py [hunk 9]
```diff
@@ -15,6 +15,7 @@
 
 KUBECTL_COMMAND = 'kubectl'
 
+
```
Comment (andrewvinall) [inline cicd/forgeops-tests/lib/utils/kubectl.py]: PEP 8:
Method definitions inside a class are surrounded by a single blank line.
Extra blank lines may be used (sparingly) to separate groups of related functions

File: cicd/forgeops-tests/lib/utils/am_pod.py [hunk 10]
```diff
@@ -25,8 +25,7 @@ def __init__(self, name, manifest_filepath):
         :param name: Pod name
         :param manifest_filepath: Path to product manifest file
         """
-
-        Pod.__init__(self, AMPod.PRODUCT_TYPE, name, manifest_filepath)
+        super().__init__(AMPod.PRODUCT_TYPE, name, manifest_filepath)
```
Comment (andrewvinall) [inline cicd/forgeops-tests/lib/utils/am_pod.py]: This change is also required in amster_pod.py for consistency.

File: cicd/forgeops-tests/lib/utils/am_pod.py [hunk 11]
```diff
@@ -55,45 +54,3 @@ def is_expected_version(self):
         assert response.json()['revision'] == self.manifest['revision'], 'Unexpected AM build revision'
         assert response.json()['date'] == self.manifest['date'], 'Unexpected AM build date'
         return True
-
```
Comment (andrewvinall) [inline cicd/forgeops-tests/lib/utils/am_pod.py]: I don't understand why these methods have been removed.

File: samples/config/smoke-deployment/env.sh (hunk 1 above)
Comment (csovant) [inline samples/config/smoke-deployment/env.sh]: Value to revert

File: cicd/forgeops-tests/tests/metadata/test_idm_metadata.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+
+        logger.test_step('Get pods')
+
+        podnames = kubectl.get_product_pod_names(TestIDMMetadata.NAMESPACE, IDMPod.PRODUCT_TYPE)
+        assert len(podnames) > 0,  'There are no AM pods'
+        for podname in podnames:
+            TestIDMMetadata.pods.append(IDMPod(podname, TestIDMMetadata.MANIFEST_FILEPATH))
+
+    def test_idm_version_info(self):
+        """Check the AM and Amster versions match"""
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/metadata/test_idm_metadata.py]: Incorrect docstring.

File: cicd/forgeops-tests/tests/metadata/test_idm_metadata.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+        """Populate the lists of pods"""
+
+        logger.test_step('Get pods')
+
+        podnames = kubectl.get_product_pod_names(TestIDMMetadata.NAMESPACE, IDMPod.PRODUCT_TYPE)
+        assert len(podnames) > 0,  'There are no AM pods'
+        for podname in podnames:
+            TestIDMMetadata.pods.append(IDMPod(podname, TestIDMMetadata.MANIFEST_FILEPATH))
+
+    def test_idm_version_info(self):
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/metadata/test_idm_metadata.py]: Maybe all products should just have a `test_version` method for consistency.

File: cicd/forgeops-tests/lib/utils/idm_pod.py [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+                                                                             % (self.manifest['date'],
+                                                                                response.json()['productBuildDate'])
+        return True
+
+    def is_expected_commons_version(self):
+        """
+        Return True if config jar is the expected version
+        :return: True if jar filename contains expected version
+        """
+        config_jar_path = os.path.join(IDMPod.ROOT, 'bundles', 'config-')
```
Comment (andrewvinall) [inline cicd/forgeops-tests/lib/utils/idm_pod.py]: This method doesn't test anything.

File: cicd/forgeops-tests/tests/metadata/test_idm_metadata.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+# Lib imports
+import os
+import pytest
+
+# Framework imports
+from utils import logger, kubectl, pod, idm_pod
+from utils.idm_pod import IDMPod
+
+
+class TestIDMMetadata(object):
```
Comment (andrewvinall) [inline cicd/forgeops-tests/tests/metadata/test_idm_metadata.py]: Do we not want to check:
* commons version
* JDK
* Legal notices

File: samples/config/smoke-deployment/env.sh (hunk 1 above)
Comment (markg58) [reply samples/config/smoke-deployment/env.sh]: Yeah, my bad.  Fixed.

[the remark this answers is not in the document]
Comment (markg58) [reply samples/config/smoke-deployment/env.sh]: yep.  Fixed

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        exec_command = [
+            'java',
+            '-version'
+        ]
+
+        resp = stream(v1.connect_get_namespaced_pod_exec, 'markg-openidm-openidm-0', 'markg',
+                      command=exec_command, container='openidm',
+                      stderr=True, stdin=True,
+                      stdout=True, tty=True, _request_timeout=30)
+
```
[the remark this answers is not in the document]
Comment (markg58) [reply cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: I should remove this file.  It was just me trying some of the various kubectl lib examples.

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+        # calling exec and wait for response.
+        exec_command = [
+            'java',
+            '-version'
+        ]
+
+        resp = stream(v1.connect_get_namespaced_pod_exec, 'markg-openidm-openidm-0', 'markg',
+                      command=exec_command, container='openidm',
+                      stderr=True, stdin=True,
+                      stdout=True, tty=True, _request_timeout=30)
```
[the remark this answers is not in the document]
Comment (markg58) [reply cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: Yes, need to make generic and not tied to my namespace.

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+        # resource_version = 'resource_version_example' # str | When specified with a watch call, shows changes that occur after that particular version of a resource. Defaults to changes from the beginning of history. When specified for list: - if unset, then the result is returned from remote storage based on quorum-read flag; - if it's 0, then we simply return what we currently have in cache, no guarantee; - if set to non zero, then the result is at least as fresh as given rv. (optional)
+        timeout_seconds = 56 # int | Timeout for the list/watch call. This limits the duration of the call, regardless of any activity or inactivity. (optional)
+
+        try:
+            api_response = v2.list_namespaced_stateful_set(namespace,
+                                                                     include_uninitialized=include_uninitialized,
+                                                                     pretty=pretty, timeout_seconds=timeout_seconds)
+            pprint(api_response)
+        except ApiException as e:
+            print("Exception when calling AppsV1beta2Api->list_namespaced_stateful_set: %s\n" % e)
```
[the remark this answers is not in the document]
Comment (markg58) [reply cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: see previous comment.

File: cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py [hunk 19]
```diff
@@ -0,0 +1,74 @@
...
+
+    def test_idm_objects(self):
+        self.list_idm_objects()
+
+    def test_list_pods(self):
```
[the remark this answers is not in the document]
Comment (markg58) [reply cicd/forgeops-tests/tests/smoke/idm/test_milestone_validation.py]: see previous comment

File: cicd/forgeops-tests/lib/utils/am_pod.py (hunk 11 above)
Comment (markg58) [reply cicd/forgeops-tests/lib/utils/am_pod.py]: Good catch.  I deleted them accidentally here.  Fixed.

File: cicd/forgeops-tests/tests/metadata/test_idm_metadata.py (hunk 12 above)
Comment (markg58) [reply cicd/forgeops-tests/tests/metadata/test_idm_metadata.py]: fixed. :thumbsup:

File: cicd/forgeops-tests/tests/metadata/test_idm_metadata.py (hunk 13 above)
Comment (markg58) [reply cicd/forgeops-tests/tests/metadata/test_idm_metadata.py]: test_version or test_version_info would probably make sense versus product specific method names for each product.

File: cicd/forgeops-tests/lib/utils/idm_pod.py (hunk 14 above)
Comment (markg58) [reply cicd/forgeops-tests/lib/utils/idm_pod.py]: method isn't finished, just wanted to get some feedback on what I had before I implemented the other methods

File: cicd/forgeops-tests/tests/metadata/test_idm_metadata.py (hunk 15 above)
Comment (markg58) [reply cicd/forgeops-tests/tests/metadata/test_idm_metadata.py]: Yes, to be implemented.  Was just looking for comments on initial code before implemented those.

File: cicd/forgeops-tests/lib/utils/idm_pod.py (hunk 14 above)
Comment (andrewvinall) [reply cicd/forgeops-tests/lib/utils/idm_pod.py]: 👍

File: cicd/forgeops-tests/tests/metadata/test_idm_metadata.py (hunk 13 above)
Comment (markg58) [reply cicd/forgeops-tests/tests/metadata/test_idm_metadata.py]: changed to test_version here

**Pull request merged**